File: BlueBlazin/lexer.py
from enum import Enum


class TokenType(Enum):
    PUNCTUATOR = 0
    IDENTIFIER = 1
    KEYWORD = 2
    NUMBER = 3
    COMMENT = 4
    EOF = 5
    BOOLEAN = 6
    STRING = 7
    NULL = 8


class Lexer:
    def __init__(self, source):
        self.source = source.rstrip()
        self.idx = 0
        self.line = 0
        self.references = {}
        self.in_reference = False

    def __iter__(self):
        tokens = list(self._lex())

        for token in tokens:
            if token["type"] == TokenType.COMMENT:
                yield from self.references[token["value"]]
            else:
                yield token

    def _lex(self):
        while self.idx < len(self.source):
            self.consume_whitespace()

            match self.peek():
                case _ if self.is_numeric():
                    yield self.scan_number()
                case "\"":
                    yield self.scan_string()
                case "+" | "-" | "*" \
                         | "(" \
                         | ")" | "{" \
                         | ";" | ":" | "[" \
                         | "]" | "," | "%":
                    yield self.token(self.line, TokenType.PUNCTUATOR, self.advance())
                case "}":
                    if self.peek(1) == "}" and self.in_reference:
                        self.in_reference = False
                        return

                    yield self.token(self.line, TokenType.PUNCTUATOR, self.advance())
                case "!" | "=" | ">" | "<":
                    match self.peek() + self.peek(1):
                        case "!=" | "==" | ">=" | "<=":
                            yield self.token(self.line, TokenType.PUNCTUATOR, self.advance(2))
                        case _:
                            yield self.token(self.line, TokenType.PUNCTUATOR, self.advance())
                case "/":
                    if self.peek(1) == "*":
                        yield from self.scan_comment()
                    else:
                        yield self.token(self.line, TokenType.PUNCTUATOR, self.advance())
                case _:
                    yield self.scan_identifier()

        yield self.token(self.line, TokenType.EOF, None)

    def scan_string(self):
        self.advance()
        line = self.line
        value = ""

        c = self.peek()

        while c != "\"":
            value += self.advance()
            c = self.peek()

        self.expect("\"")
        return self.token(line, TokenType.STRING, value)

    def scan_comment(self):
        # consume '/*'
        line = self.line
        self.advance(2)
        self.consume_whitespace()

        self.expect("[")

        if self.peek() == "[":
            self.advance()
            ref = int(self.scan_number()["value"])
            self.expect("]]")

            self.consume_whitespace()
            self.expect(r"{{")
            self.in_reference = True
            self.references[ref] = list(self._lex())
            # consume '}}'
            self.advance(2)
            self.consume_whitespace()
            self.expect("*/")
        else:
            ref = int(self.scan_number()["value"])
            self.expect("]")

            while self.idx < len(self.source):
                if self.peek() == "*" and self.peek(1) == "/":
                    self.advance(2)
                    break
                else:
                    self.advance()

            yield self.token(line, TokenType.COMMENT, ref)

    def scan_number(self):
        num = ""
        while self.is_numeric():
            num += self.advance()

        if self.peek() == ".":
            num += self.advance()
            if not ("0" <= self.peek() <= "9"):
                raise Exception(f"Syntax Error: '.' on line: {self.line}")
            else:
                while "0" <= self.peek() <= "9":
                    num += self.advance()

        return self.token(self.line, TokenType.NUMBER, float(num))

    def scan_identifier(self):
        ident = ""
        line = self.line

        if self.is_alpha() or self.peek() == "$" or self.peek() == "_":
            ident += self.advance()
        else:
            raise Exception(f"Syntax Error: Invalid identifier"
                            f" start {self.peek()} on line: {self.line}")

        while self.is_alphanumeric() or self.peek() == "$" or self.peek() == "_":
            ident += self.advance()

        if self.is_keyword(ident):
            return self.token(line, TokenType.KEYWORD, ident)
        elif ident == "true" or ident == "false":
            return self.token(line, TokenType.BOOLEAN, ident == "true")
        elif ident == "null":
            return self.token(line, TokenType.NULL, None)
        else:
            return self.token(line, TokenType.IDENTIFIER, ident)

    def is_keyword(self, ident):
        match ident:
            case "if" | "while" | "return" | "let" | "function" | "else" | "print":
                return True
            case _:
                return False

    def consume_whitespace(self):
        while self.idx < len(self.source):
            match self.peek():
                case " " | "\t" | "\n" | "\r":
                    self.advance()
                case _:
                    break

    def token(self, line, token_type, value):
        return {"type": token_type,
                "value": value,
                "line": line}

    def is_alpha(self):
        x = self.peek()

        return "a" <= x <= "z" or "A" <= x <= "Z"

    def is_numeric(self):
        return "0" <= self.peek() <= "9"

    def is_alphanumeric(self):
        return self.is_alpha() or self.is_numeric()

    def expect(self, c):
        if self.advance(len(c)) != c:
            raise Exception(
                f"Syntax Error: Unexpected character(s) or EOF on line: {self.line}")

    def peek(self, n=0):
        if self.idx + n < len(self.source):
            return self.source[self.idx + n]

        return ""

    def advance(self, n=1):
        if self.idx < len(self.source):
            value = self.source[self.idx:self.idx + n]
            self.idx += n

            if value == "\n":
                self.line += 1

            return value

        raise Exception("Syntax Error: Unexpected EOF")

File: 🌌/src/lexer.rs
use core::fmt;
use std::{iter::Peekable, str::Chars};

#[derive(Clone, Debug, PartialEq)]
pub enum TokenType {
    LeftBrace,
    RightBrace,
    LeftParen,
    RightParen,
    LeftSquare,
    RightSquare,
    Assign,
    Semicolon,
    Dot,
    Comma,
    Integer(String),
    Float(String),
    String(String),
    Identifier(String, Vec<String>),
    Func,
    Let,
    Return,
    Break,
    Continue,
    Export,
    Import,
    If,
    Else,
    While,
    Struct,
    Plus,
    Minus,
    Star,
    Slash,
    Modulo,
    And,
    Or,
    Not,
    Equal,
    NotEqual,
    LessThan,
    LessThanOrEqual,
    GreaterThan,
    GreaterThanOrEqual,
    Eof,
}

#[derive(Debug, Clone)]
pub struct Token {
    pub token_type: TokenType,
    pub line: usize,
    pub column: usize,
}

#[derive(Debug, Clone, PartialEq)]
pub struct LexerError {
    message: String,
    line: usize,
    column: usize,
}

impl fmt::Display for LexerError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(
            f,
            "{} in line {}, col {}",
            self.message, self.line, self.column
        )
    }
}

pub struct Lexer<'a> {
    chars: Peekable<Chars<'a>>,
    line: usize,
    column: usize,
}

impl<'a> Lexer<'a> {
    pub fn new(code: &'a str) -> Self {
        Self {
            chars: code.chars().peekable(),
            line: 1,
            column: 0,
        }
    }

    pub fn into_tokens(mut self) -> Result<Vec<Token>, LexerError> {
        let mut tokens = Vec::new();
        loop {
            let next = self.next()?;
            match next.token_type {
                TokenType::Eof => {
                    tokens.push(next);
                    break;
                }
                _ => {
                    tokens.push(next);
                }
            }
        }
        Ok(tokens)
    }

    fn error(&self, message: String) -> LexerError {
        LexerError {
            message,
            line: self.line,
            column: self.column,
        }
    }

    fn peek(&mut self) -> char {
        *(self.chars.peek().unwrap_or(&char::default()))
    }

    fn advance(&mut self) -> char {
        let c = self.chars.next().unwrap_or_default();
        self.column += 1;

        if c == '\n' {
            self.line += 1;
            self.column = 0;
        }

        c
    }

    fn consume(&mut self, c: char) -> Result<(), LexerError> {
        if self.peek() == c {
            self.advance();
            Ok(())
        } else {
            let found = self.peek();
            if found == char::default() {
                Err(self.error(format!("Expected '{}', found end of stream", c)))
            } else {
                Err(self.error(format!("Expected '{}', found '{}'", c, found)))
            }
        }
    }

    fn emit(&self, token_type: TokenType) -> Token {
        Token {
            token_type,
            line: self.line,
            column: self.column,
        }
    }

    pub fn next(&mut self) -> Result<Token, LexerError> {
        while self.peek().is_whitespace() {
            self.advance();
        }

        if self.peek() == '{' {
            self.advance();
            Ok(self.emit(TokenType::LeftBrace))
        } else if self.peek() == '}' {
            self.advance();
            Ok(self.emit(TokenType::RightBrace))
        } else if self.peek() == '(' {
            self.advance();
            Ok(self.emit(TokenType::LeftParen))
        } else if self.peek() == ')' {
            self.advance();
            Ok(self.emit(TokenType::RightParen))
        } else if self.peek() == '[' {
            self.advance();
            Ok(self.emit(TokenType::LeftSquare))
        } else if self.peek() == ']' {
            self.advance();
            Ok(self.emit(TokenType::RightSquare))
        } else if self.peek() == ';' {
            self.advance();
            Ok(self.emit(TokenType::Semicolon))
        } else if self.peek() == ':' {
            self.advance();
            self.consume('=')?;
            Ok(self.emit(TokenType::Assign))
        } else if self.peek() == '.' {
            self.advance();
            Ok(self.emit(TokenType::Dot))
        } else if self.peek() == ',' {
            self.advance();
            Ok(self.emit(TokenType::Comma))
        } else if self.peek() == '+' {
            self.advance();
            Ok(self.emit(TokenType::Plus))
        } else if self.peek() == '-' {
            self.advance();
            Ok(self.emit(TokenType::Minus))
        } else if self.peek() == '*' {
            self.advance();
            Ok(self.emit(TokenType::Star))
        } else if self.peek() == '/' {
            self.advance();
            Ok(self.emit(TokenType::Slash))
        } else if self.peek() == '%' {
            self.advance();
            Ok(self.emit(TokenType::Modulo))
        } else if self.peek() == '&' {
            self.advance();
            self.consume('&')?;
            Ok(self.emit(TokenType::And))
        } else if self.peek() == '|' {
            self.advance();
            self.consume('|')?;
            Ok(self.emit(TokenType::Or))
        } else if self.peek() == '!' {
            self.advance();
            if self.peek() == '=' {
                self.advance();
                Ok(self.emit(TokenType::NotEqual))
            } else {
                Ok(self.emit(TokenType::Not))
            }
        } else if self.peek() == '=' {
            self.advance();
            if self.peek() == '=' {
                self.advance();
                Ok(self.emit(TokenType::Equal))
            } else {
                Err(self.error("Invalid '=' token. Did you mean ':='?".to_owned()))
            }
        } else if self.peek() == '<' {
            self.advance();
            if self.peek() == '=' {
                self.advance();
                Ok(self.emit(TokenType::LessThanOrEqual))
            } else {
                Ok(self.emit(TokenType::LessThan))
            }
        } else if self.peek() == '>' {
            self.advance();
            if self.peek() == '=' {
                self.advance();
                Ok(self.emit(TokenType::GreaterThanOrEqual))
            } else {
                Ok(self.emit(TokenType::GreaterThan))
            }
        } else if self.peek() == '"' {
            Ok(self.consume_string()?)
        } else if self.peek().is_digit(10) {
            Ok(self.consume_number()?)
        } else if self.peek() == '`' {
            Ok(self.consume_backtick_identifier())
        } else if Self::is_naked_identifier_char(self.peek()) {
            Ok(self.consume_naked_identifier())
        } else if self.peek() == char::default() {
            Ok(self.emit(TokenType::Eof))
        } else {
            let peek = self.peek();
            Err(self.error(format!("Unexpected {}", peek)))
        }
    }

    fn consume_number(&mut self) -> Result<Token, LexerError> {
        let mut value = String::new();
        while self.peek().is_digit(10) {
            value.push(self.advance());
        }

        if self.peek() == '.' {
            self.advance();
            if !self.peek().is_digit(10) {
                return Err(self.error("Expected digit".to_owned()));
            }
            value.push('.');
            while self.peek().is_digit(10) {
                value.push(self.advance());
            }
            Ok(self.emit(TokenType::Float(value)))
        } else {
            Ok(self.emit(TokenType::Integer(value)))
        }
    }

    fn consume_string(&mut self) -> Result<Token, LexerError> {
        let mut value = String::new();
        self.consume('"')?;
        while self.peek() != '"' && self.peek() != char::default() {
            if self.peek() == '\\' {
                self.advance();
                match self.peek() {
                    'n' => {
                        value.push('\n');
                        self.advance();
                    }
                    't' => {
                        value.push('\t');
                        self.advance();
                    }
                    '"' | '\\' => {
                        value.push(self.advance());
                    }
                    c => return Err(self.error(format!("Unknown escape '\\{}'", c))),
                };
            } else {
                value.push(self.advance());
            }
        }
        self.consume('"')?;
        Ok(self.emit(TokenType::String(value)))
    }

    fn is_operator_char(c: char) -> bool {
        "{}()[]:;.,+-*/%&|<>=!".contains(c)
    }

    fn is_naked_identifier_char(c: char) -> bool {
        c != char::default() && !Self::is_operator_char(c)
    }

    fn keyword(&self, value: &str) -> Option<Token> {
        match value.trim_end() {
            "return" => Some(self.emit(TokenType::Return)),
            "let" => Some(self.emit(TokenType::Let)),
            "func" => Some(self.emit(TokenType::Func)),
            "if" => Some(self.emit(TokenType::If)),
            "else" => Some(self.emit(TokenType::Else)),
            "while" => Some(self.emit(TokenType::While)),
            "break" => Some(self.emit(TokenType::Break)),
            "continue" => Some(self.emit(TokenType::Continue)),
            "export" => Some(self.emit(TokenType::Export)),
            "import" => Some(self.emit(TokenType::Import)),
            "struct" => Some(self.emit(TokenType::Struct)),
            _ => None,
        }
    }

    fn consume_naked_identifier(&mut self) -> Token {
        let mut value = String::new();
        let mut args = Vec::new();

        while Self::is_naked_identifier_char(self.peek()) {
            if self.peek().is_whitespace() {
                if let Some(keyword_token) = self.keyword(&value) {
                    return keyword_token;
                }
            } else if self.peek() == '`' {
                let mut arg = String::new();

                value.push(self.advance());
                while self.peek() != '`' {
                    arg.push(self.advance());
                }

                value.push_str(&arg);
                args.push(arg);
            }

            value.push(self.advance())
        }

        if let Some(keyword_token) = self.keyword(value.trim_end()) {
            keyword_token
        } else {
            self.emit(TokenType::Identifier(value.trim_end().to_string(), args))
        }
    }

    fn consume_backtick_identifier(&mut self) -> Token {
        let mut value = String::new();
        self.advance();
        while self.peek() != '`' {
            value.push(self.advance())
        }
        self.advance();
        self.emit(TokenType::Identifier(value, vec![]))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn lex(code: &str) -> Result<Vec<TokenType>, LexerError> {
        Lexer::new(code)
            .into_tokens()
            .map(|tokens| tokens.into_iter().map(|t| t.token_type).collect())
    }

    #[test]
    fn a_whole_program_can_be_one_identifier() {
        let program = "why did the chicken cross the road? to get to the 0ther side";

        assert_eq!(
            lex(program),
            Ok(vec![
                TokenType::Identifier(program.to_owned(), Vec::new()),
                TokenType::Eof
            ])
        );
    }

    #[test]
    fn naked_identifier_includes_whitespace_after_it() {
        let program = "Hello world := 5;";

        assert_eq!(
            lex(program),
            Ok(vec![
                TokenType::Identifier("Hello world".to_owned(), Vec::new()),
                TokenType::Assign,
                TokenType::Integer("5".to_owned()),
                TokenType::Semicolon,
                TokenType::Eof
            ])
        );
    }

    #[test]
    fn marked_identifier_can_contain_any_characters() {
        let program = "`Hello, world! := 5 - 2 < 3` := 5;";

        assert_eq!(
            lex(program),
            Ok(vec![
                TokenType::Identifier("Hello, world! := 5 - 2 < 3".to_owned(), Vec::new()),
                TokenType::Assign,
                TokenType::Integer("5".to_owned()),
                TokenType::Semicolon,
                TokenType::Eof
            ])
        );
    }

    #[test]
    fn keyword_at_the_beginning_of_naked_identifier_is_a_keyword() {
        let program = "return hello world;";

        assert_eq!(
            lex(program),
            Ok(vec![
                TokenType::Return,
                TokenType::Identifier("hello world".to_owned(), Vec::new()),
                TokenType::Semicolon,
                TokenType::Eof
            ])
        );
    }

    #[test]
    fn keyword_within_identifier_is_part_of_identifier() {
        let program = "hello return world;";

        assert_eq!(
            lex(program),
            Ok(vec![
                TokenType::Identifier("hello return world".to_owned(), Vec::new()),
                TokenType::Semicolon,
                TokenType::Eof
            ])
        );
    }

    #[test]
    fn keyword_at_the_beginning_of_marked_identifier_is_part_of_identifier() {
        let program = "`return hello world!`;";

        assert_eq!(
            lex(program),
            Ok(vec![
                TokenType::Identifier("return hello world!".to_owned(), Vec::new()),
                TokenType::Semicolon,
                TokenType::Eof
            ])
        );
    }

    #[test]
    fn naked_identifier_can_contain_embedded_marked_identifiers_as_parameters() {
        let program = "func contains substring checks if `haystack` contains `needle` { }";

        assert_eq!(
            lex(program),
            Ok(vec![
                TokenType::Func,
                TokenType::Identifier(
                    "contains substring checks if `haystack` contains `needle`".to_owned(),
                    vec!["haystack".to_owned(), "needle".to_owned()]
                ),
                TokenType::LeftBrace,
                TokenType::RightBrace,
                TokenType::Eof
            ])
        );
    }

    #[test]
    fn operators_break_up_naked_identifiers() {
        let program = "let Hello, world := 42;";

        assert_eq!(
            lex(program),
            Ok(vec![
                TokenType::Let,
                TokenType::Identifier("Hello".to_owned(), Vec::new()),
                TokenType::Comma,
                TokenType::Identifier("world".to_owned(), Vec::new()),
                TokenType::Assign,
                TokenType::Integer("42".to_owned()),
                TokenType::Semicolon,
                TokenType::Eof
            ])
        );
    }

    #[test]
    fn operators_do_not_break_up_marked_identifiers() {
        let program = "let `Hello, world!` := 42;";

        assert_eq!(
            lex(program),
            Ok(vec![
                TokenType::Let,
                TokenType::Identifier("Hello, world!".to_owned(), Vec::new()),
                TokenType::Assign,
                TokenType::Integer("42".to_owned()),
                TokenType::Semicolon,
                TokenType::Eof
            ])
        );
    }

    #[test]
    fn string_without_escape_characters() {
        let program = "print with new line(\"Hello, world!\");";

        assert_eq!(
            lex(program),
            Ok(vec![
                TokenType::Identifier("print with new line".to_owned(), Vec::new()),
                TokenType::LeftParen,
                TokenType::String("Hello, world!".to_owned()),
                TokenType::RightParen,
                TokenType::Semicolon,
                TokenType::Eof
            ])
        );
    }

    #[test]
    fn string_with_escape_characters() {
        let program = "\"Hello,\\nworld\\t!\\\\\\\"\"";
        assert_eq!(
            lex(program),
            Ok(vec![
                TokenType::String("Hello,\nworld\t!\\\"".to_owned()),
                TokenType::Eof
            ])
        );
    }

    #[test]
    fn string_with_invalid_escape_characters_causes_error() {
        let program = "\"Invalid \\escape\"";
        assert_eq!(
            lex(program),
            Err(LexerError {
                message: "Unknown escape '\\e'".to_owned(),
                line: 1,
                column: 10,
            })
        );
    }

    #[test]
    fn unterminated_string() {
        let program = "\"Hello, world!";
        assert_eq!(
            lex(program),
            Err(LexerError {
                message: "Expected '\"', found end of stream".to_owned(),
                line: 1,
                column: 14,
            })
        );
    }

    #[test]
    fn malformed_operators_return_errors() {
        let program = "let x : 2;";

        assert_eq!(
            lex(program),
            Err(LexerError {
                message: "Expected '=', found ' '".to_owned(),
                line: 1,
                column: 7
            })
        );
    }

    #[test]
    fn lexer_reports_location_of_error() {
        let program = "let x := 2;\nlet y := 3;\n\nlet z := 1 | 2;\n\nlet w := 7;";

        assert_eq!(
            lex(program),
            Err(LexerError {
                message: "Expected '|', found ' '".to_owned(),
                line: 4,
                column: 12
            })
        );
    }
}
